articles/realtime.md

```diff
@@ -0,0 +1,330 @@
+author: Jackie Kay
+---
+
+* This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+This document seeks to summarize the requirements of real-time computing and the challenges of implementing real-time performance.
+It also lays out options for how ROS 2 could be structured to enforce real-time compatibility.
+
+Robotic systems need to be responsive.
+In mission critical applications, a delay of less than a millisecond in the system can cause a catastrophic failure.
+For ROS 2.0 to capture the needs of the robotics community, the core software components must not interfere with the requirements of real-time computing.
+
+# Definition of Real-time Computing
+
+The definition of real-time computing requires the definition of a few other key terms:
+
+* Determinism: A system is deterministic if it always produces the same output for a known input. The output of a nondeterministic system will have random variations.
+* Deadline: A deadline is the finite window of time in which a certain task must be completed.
+* Quality of Service: The overall performance of a network. Includes factors such as bandwith, throughput, availability, jitter, latency, and error rates.
+
+Real-time software guarantees correct computation at the correct time.
+Hard real-time software systems have a set of strict deadlines, and missing a deadline is considered a system failure.
+Examples of hard real-time systems: airplane sensor and autopilot systems, spacecrafts and planetary rovers.
+
+Soft real-time systems try to reach deadlines but do not fail if a deadline is missed.
+However, they may degrade their quality of service in such an event to improve responsiveness.
+Examples of soft real-time systems: audio and video delivery software for entertainment (lag is undesirable but not catastrophic).
+
+Firm real-time systems treat information delivered/computations made after a deadline as invalid.
+Like soft real-time systems, they do not fail after a missed deadline, and they may degrade QoS if a deadline is missed. (1)
+Examples of firm real-time systems: financial forecast systems, robotic assembly lines. (2)
+
+Real-time computer systems are often associated with low-latency systems.
+Many applications of real-time computing are also low-latency applications (for example, automated piloting systems must be reactive to sudden changes in the environment).
+However, a real-time system is not defined by low latency, but by a deterministic schedule: it must be guaranteed that the system finishes a certain task by a certain time.
+Therefore, it is important that the latency in the system be measurable and a maximum allowable latency for tasks be set.
+
+A real-time computer system needs both an operating system that can handle scheduling and memory allocation in real-time and user code that delivers deterministic execution.
```

Comment:
I don't think it's necessary to explicitly mention memory allocation here. I would say that handling memory allocations in a real-time safe manner can be considered part of writing real-time safe code.